sdk/compiler/daml-lf-tools/src/DA/Daml/LF/TypeChecker/Upgrade.hs

```diff
@@ -812,79 +833,129 @@ checkType type_ err = do
+                    _ -> [] -- This case should not be possible
+                Upgrading False False ->
+                  case udMbPackageVersion <$> upgradingDep of
+                      Upgrading (Just pastVersion) (Just presentVersion) -> ifMismatch (pastVersion < presentVersion) ("Name came from package '" <> show (_past upgradingDep) <> "' and now comes from package '" <> show (_present upgradingDep) <> "'. Both packages support upgrades, but the previous package had a higher version than the current one.")
+                      Upgrading Nothing Nothing -> [] -- Only daml-prim has no version field
+                      _ -> [] -- This case should not be possible
+                Upgrading False True -> [nameMismatch' ("Name came from package '" <> show (_past upgradingDep) <> "' and now comes from package '" <> show (_present upgradingDep) <> "'. Both packages support upgrades, but the previous package was not a utility package and the current one is.")]
+                Upgrading True False -> [nameMismatch' ("Name came from package '" <> show (_past upgradingDep) <> "' and now comes from package '" <> show (_present upgradingDep) <> "'. Both packages support upgrades, but the previous package was a utility package and the current one is not.")]
+            Upgrading False True -> [nameMismatch' ("Name came from package '" <> show (_past upgradingDep) <> "' and now comes from package '" <> show (_present upgradingDep) <> "'. The previous package did not support upgrades and the current one does.")]
+            Upgrading True False -> [nameMismatch' ("Name came from package '" <> show (_past upgradingDep) <> "' and now comes from package '" <> show (_present upgradingDep) <> "'. The previous package supported upgrades and the current one does not.")]
+        Upgrading { _past = Nothing, _present = Nothing } ->
+          [nameMismatch' ("Name came from package '" <> tryShowPkgId (_past pkgId) <> "' and now comes from package '" <> tryShowPkgId (_present pkgId) <> "'. Could not find " <> tryShowPkgId (_past pkgId) <> " nor " <> tryShowPkgId (_present pkgId) <> " in the package list.")]
+        Upgrading { _past = Just _, _present = Nothing } ->
+          [nameMismatch' ("Name came from package '" <> tryShowPkgId (_past pkgId) <> "' and now comes from package '" <> tryShowPkgId (_present pkgId) <> "'. Could not find " <> tryShowPkgId (_present pkgId) <> " in the package list.")]
+        Upgrading { _past = Nothing, _present = Just _ } ->
+          [nameMismatch' ("Name came from package '" <> tryShowPkgId (_past pkgId) <> "' and now comes from package '" <> tryShowPkgId (_present pkgId) <> "'. Could not find " <> tryShowPkgId (_past pkgId) <> " in the package list.")]
+
+checkQualType :: UpgradingDeps -> Upgrading (Qualified TypeConName) -> [Mismatch]
```

Comment:
Given the size of these functions now, it probably makes sense to try dedup. The `checkQualType` differs in that it doesn't check utility packages, but even if it had the code to do so, it'd never hit that branch. Could you not simply use the same (ish) function for both?

Reply:
The main issue is printing different messages for types vs values. For values, the message is a warning, and worded as such. For types, not so much.